app/src/components/validators.js

```diff
@@ -0,0 +1,315 @@
+const bytes = require('bytes');
+const fs = require('fs');
+const log = require('npmlog');
+const tmp = require('tmp');
+const validator = require('validator');
+
+const DEFAULT_ATTACHMENT_SIZE = bytes.parse('5mb');
+
+const asyncForEach = async (array, callback) => {
+  for (let index = 0; index < array.length; index++) {
+    await callback(array[index], index, array);
+  }
+};
+
+const validators = {
+
+  attachment: {
+    content: value => {
+      return validatorUtils.isString(value) && !validator.isEmpty(value, {ignore_whitespace: true});
+    },
+
+    contentType: value => {
+      if (value) {
+        return validatorUtils.isString(value) && !validator.isEmpty(value, {ignore_whitespace: true});
+      }
+      return true;
+    },
+
+    encoding: value => {
+      if (value) {
+        return validatorUtils.isString(value) && !validator.isEmpty(value, {ignore_whitespace: true}) && validator.isIn(value, ['base64','binary','hex']);
+      }
+      return true;
+    },
+
+    filename: value => {
+      return validatorUtils.isString(value) && !validator.isEmpty(value, {ignore_whitespace: true});
+    },
+
+    size: async (content, encoding, limit = DEFAULT_ATTACHMENT_SIZE) => {
+      if (!(validators.attachment.content(content) && validators.attachment.encoding(encoding))) {
+        return false;
+      }
+
+      let attachmentLimit = bytes.parse(limit);
+      if (!attachmentLimit || isNaN(attachmentLimit)|| attachmentLimit < 1) {
+        return false;
+      }
+
+      // ok, looks like all incoming parameters are ok, check the size
+      // write out temp file, if size is ok then return true...
+      let tmpFile = undefined;
+
+      try {
+        tmpFile = tmp.fileSync();
+        await fs.promises.writeFile(tmpFile.name, new Buffer(content, encoding));
+        // get the written file size
+        const stats = fs.statSync(tmpFile.name);
+        return stats.size <= attachmentLimit;
+      } catch(e) {
+        // something wrong (disk i/o?), cannot verify file size
+        log.error(`Error validating file size. ${e.message}`);
+        return false;
+      } finally {
+        // delete tmp file
+        if (tmpFile) tmpFile.removeCallback();
+      }
+
+    }
+  },
+
+  attachments: async (obj, attachmentSizeLimit = DEFAULT_ATTACHMENT_SIZE) => {
```

Comment:
We should be adding in JSDoc annotations to these top-level functions whenever possible.

Reply:
save for a cleanup later